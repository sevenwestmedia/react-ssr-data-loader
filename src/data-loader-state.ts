export interface SuccessAction {
    type: 'none' | 'fetch'
    success: true
}

export interface FailedAction {
    type: 'fetch'
    success: false
    error: Error
}

export const enum LoaderStatus {
    /**
     * The loader is inactive -- not performing any action
     */
    Idle = 'Idle',

    /**
     * The loader is fetching the resource
     */
    Fetching = 'Fetching',
}

export interface LoaderState<TData> {
    status: LoaderStatus
    lastAction: SuccessAction | FailedAction
    data: Data<TData>
}

// @ TODO Should we drop dataFromServerSideRender? How do we model not fetching on client
export type Data<TData> =
    | {
          hasData: true
          result: TData
          dataFromServerSideRender: boolean
      }
    | { hasData: false }
