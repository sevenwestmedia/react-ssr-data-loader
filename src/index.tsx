export * from './data-loader-resources'
export { DataLoaderActions } from './data-loader'
export { DataProvider, DataProviderProps } from './data-provider'
export { LoaderState, LoaderStatus } from './data-loader-state'
export { DataProviderEvents } from './events'
export { DataLoaderState } from './data-loader-store-and-loader'
