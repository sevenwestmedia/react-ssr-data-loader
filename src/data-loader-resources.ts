import { createTypedDataLoader, DataLoaderActions } from './data-loader'
import { LoaderStatus } from './data-loader-state'

export type LoadResource<TData, TResourceParameters, TInternalState, TGlobalParameters> = (
    params: { resourceType: string } & TResourceParameters & TInternalState & TGlobalParameters,
) => Promise<TData> | TData

interface Resources {
    [dataType: string]: LoadResource<any, any, any, any>
}

export interface PagedData<Data> {
    pageNumber: number
    data: Data
}

export interface Paging {
    /** Defaults to 0 */
    initialOffset?: number
    /** Overrides pageSize for initial page */
    initialSize?: number
    pageSize: number
}
export interface PageComponentProps {
    paging: Paging
}
export interface PageState {
    page: number
}

/** TGlobalParameters is provided through the data provider, and accessible in all data load function */
export class DataLoaderResources<TGlobalParameters> {
    private resources: Resources = {}

    registerResource<TData, TResourceParameters>(
        resourceType: string,
        loadResource: LoadResource<TData, TResourceParameters, {}, TGlobalParameters>,
        cacheKeyProperties?: Array<keyof TResourceParameters & string>,
    ) {
        if (this.resources[resourceType]) {
            throw new Error(`The resource type ${resourceType} has already been registered`)
        }

        const actions: DataLoaderActions<{}, TData> = {
            refresh(internalState) {
                return {
                    newInternalState: internalState,
                    refresh: true,
                    keepData: true,
                }
            },
        }
        const typedDataLoader = createTypedDataLoader<
            TData,
            TResourceParameters,
            {},
            typeof actions
        >(resourceType, {}, actions, cacheKeyProperties)
        this.resources[resourceType] = loadResource

        return typedDataLoader
    }

    /** Page numbers start at 1 */
    registerPagedResource<TData, TResourceParameters>(
        resourceType: string,
        loadResource: LoadResource<
            TData,
            TResourceParameters & PageComponentProps,
            PageState,
            TGlobalParameters
        >,
        cacheKeyProperties?: Array<keyof TResourceParameters & string>,
    ) {
        if (this.resources[resourceType]) {
            throw new Error(`The resource type ${resourceType} has already been registered`)
        }

        const actions: DataLoaderActions<PageState, PagedData<TData>> = {
            refresh(internalState) {
                return {
                    newInternalState: internalState,
                    refresh: true,
                    keepData: true,
                }
            },
            // Loads next page of data
            nextPage(internalState, state) {
                if (state.status !== LoaderStatus.Idle) {
                    return null
                }
                return {
                    keepData: false,
                    refresh: false,
                    newInternalState: { page: internalState.page + 1 },
                }
            },
        }

        const typedDataLoader = createTypedDataLoader<
            PagedData<TData>,
            PageComponentProps & TResourceParameters,
            PageState,
            typeof actions
        >(resourceType, { page: 1 }, actions, cacheKeyProperties)
        this.resources[resourceType] = loadResource

        return typedDataLoader
    }

    getResourceLoader(dataType: any): LoadResource<any, any, any, any> {
        return this.resources[dataType]
    }
}
