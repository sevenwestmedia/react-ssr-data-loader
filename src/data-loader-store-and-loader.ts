import isPromise from 'is-promise'
import { DataProviderEvents } from './events'
import { LoaderState, LoaderStatus } from './data-loader-state'
import { getDataState } from './state-helper'

export interface DataLoaderState {
    [paramsHash: string]: LoaderState<any>
}

export interface ActionResult<TInternalState> {
    /** Directive to keep the existing data if this action causes a data load */
    keepData: boolean
    /** Directive to refresh the data, even if the inputs are all the same */
    refresh: boolean
    /** Internal state for this data loader, useful for paging and other exensions */
    newInternalState: TInternalState
}

export interface LoadParams {
    resourceType: string
    [param: string]: any
}

/** Takes an object and produces a consistent hash */
export type ObjectHash = (obj: object) => string

// Some other names
// DataLoaderDataAccessor
// DataLoaderCore
// DataLoaderSmurf
// ?
// tslint:disable-next-line:max-classes-per-file
export class DataLoaderStoreAndLoader {
    private requiresUpdateOnMount: string[] = []

    private registeredDataLoaders: {
        [dataLoaderId: string]: {
            update: () => void
            currentParamsHash: string | undefined
        }
    } = {}

    /** Lookup of the dataLoaderIds which currently are consuming the params. Used for ref counting */
    private paramHashConsumers: { [paramsHash: string]: string[] } = {}

    // Version allows us to throw away events which are not for the correct version
    // for example, refresh while loading initial data. We only want final refresh
    // result
    private paramHashVersion: { [paramsHash: string]: number } = {}
    private dataStore: DataLoaderState = {}
    private stateVersionCounter = 0
    private currentWorkCount = 0
    private onEvent: (event: DataProviderEvents) => void | Promise<any>

    constructor(
        onEvent: (event: DataProviderEvents) => void | Promise<any>,
        initialState: DataLoaderState | undefined,
        private performLoad: (dataLoadParams: LoadParams) => Promise<any> | any,
        private objectHash: ObjectHash,
        public isServerSideRender: boolean,
    ) {
        this.onEvent = event => {
            try {
                return onEvent(event)
            } catch (err) {
                // eslint-disable-next-line no-console
                console.error('onEvent handler threw', err)
            }
        }
        if (initialState) {
            this.dataStore = initialState
        } else {
            // Raise state change only if we haven't got initial state
            this.onEvent({
                type: 'state-changed',
                state: { ...this.dataStore },
            })
        }
    }

    createHash(dataLoadParams: object, cacheKeyProperties: string[] | undefined) {
        const cacheParams = cacheKeyProperties
            ? // ensure resourceType is always included in the cache key
              cacheKeyProperties.concat('resourceType').reduce<any>((acc, val) => {
                  acc[val] = (dataLoadParams as any)[val]
                  return acc
              }, {})
            : dataLoadParams

        return this.objectHash(cacheParams)
    }

    attach(
        componentInstanceId: string,
        resourceType: string,
        dataLoadParams: object,
        cacheKeyProperties: string[] | undefined,
        update: () => void,
    ): void {
        const paramsObject = this.getParamsObject(resourceType, dataLoadParams)
        const paramsObjectHash = this.createHash(paramsObject, cacheKeyProperties)

        this.registerDataLoaderHash(paramsObjectHash, componentInstanceId)
        this.registeredDataLoaders[componentInstanceId] = {
            update,
            currentParamsHash: paramsObjectHash,
        }

        // If the data has resolved between initial render, and mounting
        // then we need to dispatch an update if not in a fetching state
        const requiresUpdateIndex = this.requiresUpdateOnMount.indexOf(componentInstanceId)
        if (requiresUpdateIndex !== -1) {
            this.requiresUpdateOnMount.splice(requiresUpdateIndex, 1)
            if (this.dataStore[paramsObjectHash].status !== LoaderStatus.Fetching) {
                update()
            }
        }
    }

    // Returns true when data needs to be unloaded from redux
    detach(
        componentInstanceId: string,
        resourceType: string,
        dataLoadParams: object,
        cacheKeyProperties: string[] | undefined,
    ) {
        const paramsObject = this.getParamsObject(resourceType, dataLoadParams)
        const paramsObjectHash = this.createHash(paramsObject, cacheKeyProperties)
        this.cleanupDataLoader(paramsObjectHash, componentInstanceId)
        delete this.registeredDataLoaders[componentInstanceId]
    }

    /** Should only be called from render of data loader component (through context) */
    getDataLoaderState(
        componentInstanceId: string,
        resourceType: string,
        dataLoadParams: object,
        cacheKeyProperties: string[] | undefined,
        keepData = false,
        forceRefresh = false,
    ) {
        const paramsObject = this.getParamsObject(resourceType, dataLoadParams)

        // Initial render (before mount) and SSR will not be registered, we can safely fall back to
        // undefined in both these instances
        const previousRenderParamsObjectHash = this.registeredDataLoaders[componentInstanceId]
            ? this.registeredDataLoaders[componentInstanceId].currentParamsHash
            : undefined
        const paramsObjectHash = this.createHash(paramsObject, cacheKeyProperties)
        const stateForParams = this.dataStore[paramsObjectHash]

        // Cleanup data which is not needed
        if (
            previousRenderParamsObjectHash &&
            paramsObjectHash !== previousRenderParamsObjectHash &&
            !keepData
        ) {
            this.cleanupDataLoader(previousRenderParamsObjectHash, componentInstanceId)
        }

        // If we have state for the current renders params, return it synchronously
        if (stateForParams && !forceRefresh) {
            return stateForParams
        }

        // Register the calling data loader as a consumer
        const wasRegistered = this.registerDataLoaderHash(paramsObjectHash, componentInstanceId)

        let result: any
        try {
            result = this.performLoad(paramsObject)
        } catch (err) {
            this.updateParamsHashState(paramsObjectHash, {
                status: LoaderStatus.Idle,
                lastAction: {
                    type: 'fetch',
                    success: false,
                    error: err,
                },
                data: getDataState(true, previousRenderParamsObjectHash, this.dataStore),
            })

            return this.dataStore[paramsObjectHash]
        }

        if (isPromise(result)) {
            if (wasRegistered) {
                // If this render caused the data load and the data
                // loader was registered, we need to register it
                this.requiresUpdateOnMount.push(componentInstanceId)
            }
            // Init state as loading
            this.updateParamsHashState(paramsObjectHash, {
                status: LoaderStatus.Fetching,
                lastAction: {
                    type: 'none',
                    success: true,
                },
                data: getDataState(keepData, previousRenderParamsObjectHash, this.dataStore),
            })
            this.onEvent({
                type: 'begin-loading-event',
                data: {
                    resourceLoadParamsHash: paramsObjectHash,
                    resourceType,
                },
            })
            this.monitorLoad(resourceType, paramsObjectHash, result, this.stateVersionCounter)
        } else {
            // Init state as loaded, nothing async to monitor
            this.updateParamsHashState(paramsObjectHash, {
                status: LoaderStatus.Idle,
                lastAction: {
                    type: 'fetch',
                    success: true,
                },
                data: {
                    hasData: true,
                    dataFromServerSideRender: this.isServerSideRender,
                    result,
                },
            })
        }

        return this.dataStore[paramsObjectHash]
    }

    private getParamsObject(resourceType: string, dataLoadParams: object) {
        return {
            resourceType,
            ...dataLoadParams,
        }
    }

    private monitorLoad(
        resourceType: string,
        paramsObjectHash: string,
        work: PromiseLike<any>,
        beginLoadVersion: number,
    ) {
        this.currentWorkCount++
        work.then(
            result => ({ success: true as true, result }),
            err => ({ success: false as false, err }),
        ).then(result => {
            const currentState = this.dataStore[paramsObjectHash]
            const currentStateVersion = this.paramHashVersion[paramsObjectHash]
            const versionMismatch = currentState && currentStateVersion !== beginLoadVersion

            const dataLoadersForHash = this.paramHashConsumers[paramsObjectHash]

            // We shouldn't update internal state if there is a version
            // mismatch, or there are no data loaders left who care
            // about this data
            if (!versionMismatch && dataLoadersForHash) {
                if (result.success) {
                    this.updateParamsHashState(paramsObjectHash, {
                        status: LoaderStatus.Idle,
                        lastAction: {
                            type: 'fetch',
                            success: true,
                        },
                        data: {
                            hasData: true,
                            dataFromServerSideRender: this.isServerSideRender,
                            result: result.result,
                        },
                    })
                } else {
                    this.updateParamsHashState(paramsObjectHash, {
                        status: LoaderStatus.Idle,
                        lastAction: {
                            type: 'fetch',
                            success: false,
                            error: result.err,
                        },
                        data: currentState.data,
                    })
                }

                dataLoadersForHash.forEach(dataLoader => {
                    const dataLoaderInfo = this.registeredDataLoaders[dataLoader]
                    // If data loaders are not mounted yet, can't call update..
                    // TODO when it registers, if the state has changed, need to call update
                    // Otherwise it will be out of sync
                    if (dataLoaderInfo) {
                        // Tell the data loader to update
                        dataLoaderInfo.update()
                    }
                })
            }

            if (!result.success) {
                this.onEvent({
                    type: 'load-error',
                    data: {
                        resourceType,
                        resourceLoadParamsHash: paramsObjectHash,
                        error: result.err,
                    },
                })
            }

            this.onEvent({
                type: 'end-loading-event',
                data: {
                    resourceType,
                    resourceLoadParamsHash: paramsObjectHash,
                },
            })

            const currentCount = --this.currentWorkCount
            if (currentCount === 0) {
                this.onEvent({
                    type: 'data-load-completed',
                    data: {
                        resourceType,
                        resourceLoadParamsHash: paramsObjectHash,
                    },
                })
            }
        })
    }

    private registerDataLoaderHash(paramsObjectHash: string, componentInstanceId: string) {
        if (!this.paramHashConsumers[paramsObjectHash]) {
            this.paramHashConsumers[paramsObjectHash] = []
        }
        if (this.paramHashConsumers[paramsObjectHash].indexOf(componentInstanceId) === -1) {
            this.paramHashConsumers[paramsObjectHash].push(componentInstanceId)
            return true
        }

        return false
    }

    private cleanupDataLoader(previousRenderParamsObjectHash: string, componentInstanceId: string) {
        const consumersForPreviousHash = this.paramHashConsumers[previousRenderParamsObjectHash]
        if (!consumersForPreviousHash) {
            return
        }
        consumersForPreviousHash.splice(consumersForPreviousHash.indexOf(componentInstanceId), 1)
        if (consumersForPreviousHash.length === 0) {
            // We no longer need this data
            delete this.dataStore[previousRenderParamsObjectHash]
            delete this.paramHashConsumers[previousRenderParamsObjectHash]

            this.onEvent({
                type: 'state-changed',
                state: { ...this.dataStore },
            })
        }
    }

    private updateParamsHashState(paramsObjectHash: string, newState: LoaderState<any>) {
        this.dataStore[paramsObjectHash] = newState

        this.paramHashVersion[paramsObjectHash] =
            (this.paramHashVersion[paramsObjectHash] || -1) + 1
        this.onEvent({
            type: 'state-changed',
            state: { ...this.dataStore },
        })
    }
}
