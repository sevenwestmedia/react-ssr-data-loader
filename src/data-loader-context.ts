import React from 'react'
import { LoaderState } from './data-loader-state'
import { DataLoaderStoreAndLoader } from './data-loader-store-and-loader'

export const DataLoaderContextComponent = React.createContext<DataLoaderStoreAndLoader | undefined>(
    undefined,
)

export function ensureContext(
    context: DataLoaderStoreAndLoader | undefined,
): DataLoaderStoreAndLoader {
    if (!context) {
        throw new Error(
            'Data loader context missing, ensure you have wrapped your application in a DataLoaderProvider',
        )
    }

    return context
}

export function ssrNeedsData(state: LoaderState<any> | undefined): boolean {
    return !state || (!state.data.hasData && state.lastAction.success)
}
