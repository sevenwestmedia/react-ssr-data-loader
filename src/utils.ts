import { LoadParams } from './data-loader-store-and-loader'

export function getError(metadata: LoadParams, err: any, fallbackMsg?: string): Error {
    if (err instanceof Error) {
        ;(err as any).dataLoadContext = `${metadata.resourceType}`

        return err
    }
    if (typeof err === 'string') {
        const error = new Error(err)
        ;(error as any).dataLoadContext = `${metadata.resourceType}`
        return error
    }

    const fallbackEror = new Error((err || fallbackMsg).toString())
    ;(fallbackEror as any).dataLoadContext = `${metadata.resourceType}`
    return fallbackEror
}
