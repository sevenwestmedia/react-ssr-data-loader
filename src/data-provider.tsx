import React from 'react'
import { DataLoaderResources } from './data-loader-resources'
import { DataProviderEvents } from './events'
import { DataLoaderContextComponent } from './data-loader-context'
import { DataLoaderStoreAndLoader, DataLoaderState } from './data-loader-store-and-loader'

export interface Props {
    initialState?: DataLoaderState
    onEvent?: (event: DataProviderEvents) => void | Promise<any>
    isServerSideRender?: boolean
    resources: DataLoaderResources<any>
    globalProps?: object
}

export const DataLoaderProvider: React.FC<Props> = props => {
    const [dataLoader] = React.useState(() => {
        return new DataLoaderStoreAndLoader(
            // tslint:disable-next-line:no-empty
            props.onEvent || (() => {}),
            props.initialState,
            params => {
                const dataLoader = props.resources.getResourceLoader(params.resourceType)
                if (!dataLoader) {
                    return Promise.reject(`No data loader present for ${params.resourceType}`)
                }

                return dataLoader.loadResource({
                    ...props.globalProps,
                    ...params,
                })
            },
            (resourceType, dataLoadParams) => {
                return props.resources.generateCacheKey(resourceType, dataLoadParams)
            },
            props.isServerSideRender || false,
        )
    })

    return (
        <DataLoaderContextComponent.Provider value={dataLoader}>
            {props.children}
        </DataLoaderContextComponent.Provider>
    )
}
DataLoaderProvider.displayName = 'DataLoaderProvider'
