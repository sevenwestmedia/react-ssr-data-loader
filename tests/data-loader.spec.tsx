import React from 'react'
import Adapter from 'enzyme-adapter-react-16'
import { mount, configure } from 'enzyme'
import { act } from 'react-dom/test-utils'
import { ComponentFixture } from './helpers/component-fixture'
import { SharedDataComponentFixture } from './helpers/shared-data-component-fixture'
import { DifferentKeysDataComponentFixture } from './helpers/different-keys-data-component-fixture'
import { DataProvider } from '../src/data-provider'
import { DataLoaderResources } from '../src/index'
import { processEventLoop } from './helpers/event-loop-helpers'

configure({ adapter: new Adapter() })

describe('data-loader', () => {
    it('supports multiple loaders using the same key when data loading', async () => {
        const sut = new SharedDataComponentFixture(undefined, 'testKey', false)

        sut.assertState()
    })

    it('can resolve data from multiple components', async () => {
        const sut = new SharedDataComponentFixture(undefined, 'testKey', false)

        await act(() => sut.testDataPromise.resolve({ result: 'Test' }))

        sut.assertState()
    })

    it('can load multiple dataloaders with different keys', async () => {
        const sut = new DifferentKeysDataComponentFixture(undefined, 'testKey', 'testKey2', false)

        await act(() => sut.testDataPromise.resolve({ result: 'Test' }))

        sut.assertState()
    })

    it('multiple components load data once when props change', async () => {
        const sut = new SharedDataComponentFixture(undefined, 'testKey', false)

        await act(async () => {
            await sut.testDataPromise.resolve({ result: 'Success!' })
            sut.resetPromise()
            sut.root.setProps({ id: 'newData' })
            await sut.testDataPromise.resolve({ result: 'Success2!' })
        })

        sut.assertState()
    })

    it('data is not unloaded until the last attached data-loader is unmounted', async () => {
        const sut = new SharedDataComponentFixture(undefined, 'testKey', false, false, true)

        await act(async () => {
            await sut.testDataPromise.resolve({ result: 'Success!' })
            sut.root.unmount()
            await processEventLoop()
        })

        sut.assertState()
    })

    it('handles onEvent throwing gracefully', async () => {
        const errors: string[] = []
        console.error = jest.fn((...err: any[]) => {
            errors.push(JSON.stringify(err))
        })
        const sut = new ComponentFixture(undefined, 'testKey', {
            isServerSideRender: false,
            onEvent: () => {
                throw new Error('Boom')
            },
        })

        await act(async () => {
            await sut.testDataPromise.resolve({ result: 'Test' })
        })

        sut.assertState()
        expect(errors).toMatchSnapshot()
    })

    it('ignores completion if unmounted first', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: false })

        await act(async () => {
            await sut.unmount()
            await sut.testDataPromise.resolve({ result: 'Test' })
        })

        sut.assertState()
    })

    it('notifies when all work is completed', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', { isServerSideRender: false })

        expect(sut.loadAllCompletedCalled).toBe(0)
        await sut.testDataPromise.resolve({ result: 'Test' })
        expect(sut.loadAllCompletedCalled).toBe(1)
    })

    it('throws when the same resource is registered multiple times', () => {
        const resources = new DataLoaderResources()

        resources.registerResource('test', () => Promise.resolve('test'))
        expect(() => resources.registerResource('test', () => Promise.resolve('test'))).toThrow()
    })

    it('throws when the same paged resource is registered multiple times', () => {
        const resources = new DataLoaderResources()

        resources.registerResource('test', () => Promise.resolve('test'))
        expect(() => resources.registerResource('test', () => Promise.resolve('test'))).toThrow()
    })

    it('resource can resolve synchronously', async () => {
        const resources = new DataLoaderResources()

        const useLoadData = resources.registerResource<string, { id: string }>(
            'test',
            () => 'Result!',
        )
        let loadCount = 0

        const LoadTest: React.FC = () => {
            const { data } = useLoadData({ id: 'Test!' })
            loadCount++

            if (data.hasData) {
                return <div>{data.result}</div>
            }

            return <div>No data!</div>
        }
        const wrapper = mount(
            <DataProvider resources={resources}>
                <LoadTest />
            </DataProvider>,
        )

        expect(wrapper.html()).toMatchSnapshot()
        await processEventLoop()

        expect(loadCount).toBe(1)
    })

    it('data loader handles synchronous throw', async () => {
        const resources = new DataLoaderResources()

        const useLoadData = resources.registerResource<string, { id: string }>('test', () => {
            throw new Error('Synchronous fail')
        })
        let loadCount = 0

        const LoadTest: React.FC = () => {
            const { data, lastAction } = useLoadData({ id: 'Test!' })
            loadCount++
            if (!lastAction.success) {
                return <div>{lastAction.error.message}</div>
            }
            if (data.hasData) {
                return <div>{data.result}</div>
            }

            return <div>No data!</div>
        }

        const wrapper = mount(
            <DataProvider resources={resources}>
                <LoadTest />
            </DataProvider>,
        )

        expect(wrapper.html()).toMatchSnapshot()
        await processEventLoop()

        expect(loadCount).toBe(1)
    })

    it('resource can resolve synchronously when resource changes', async () => {
        const sut = new ComponentFixture(undefined, 'testKey', {
            isServerSideRender: false,
            unloadDataOnUnmount: false,
            syncResult: {
                result: 'Result!',
            },
        })

        sut.testDataResult = {
            result: 'Result2!',
        }
        sut.setId('newData')

        sut.assertState()
    })
})
