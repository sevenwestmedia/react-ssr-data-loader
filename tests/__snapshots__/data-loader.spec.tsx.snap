// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data-loader can load multiple dataloaders with different keys 1`] = `
Object {
  "loadAllCompletedCalled": 0,
  "loadDataCount1": 1,
  "loadDataCount2": 1,
  "renderCount": 2,
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Test",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
  "renderProps2": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Fetching",
  },
}
`;

exports[`data-loader can resolve data from multiple components 1`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderCount": 2,
  "renderCount2": 2,
  "renderProps1": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Test",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
  "renderProps2": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Test",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`data-loader data is not unloaded until the last attached data-loader is unmounted 1`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderCount": 2,
  "renderCount2": 2,
  "renderProps1": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Success!",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
  "renderProps2": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Success!",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`data-loader data loader handles synchronous throw 1`] = `"<div>Synchronous fail</div>"`;

exports[`data-loader handles onEvent throwing gracefully 1`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Test",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "fetch",
          },
          "status": "Idle",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "end-loading-event",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "data-load-completed",
    },
  ],
  "loadAllCompletedCalled": 0,
  "loadDataCount": 1,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 2,
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Test",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`data-loader handles onEvent throwing gracefully 2`] = `
Array [
  "[\\"onEvent handler threw\\",{}]",
  "[\\"onEvent handler threw\\",{}]",
  "[\\"onEvent handler threw\\",{}]",
  "[\\"onEvent handler threw\\",{}]",
  "[\\"onEvent handler threw\\",{}]",
  "[\\"onEvent handler threw\\",{}]",
]
`;

exports[`data-loader ignores completion if unmounted first 1`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "hasData": false,
          },
          "lastAction": Object {
            "success": true,
            "type": "none",
          },
          "status": "Fetching",
        },
      },
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "begin-loading-event",
    },
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "end-loading-event",
    },
    Object {
      "data": Object {
        "resourceLoadParamsHash": "59edd2a8",
        "resourceType": "testDataType",
      },
      "type": "data-load-completed",
    },
  ],
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 1,
  "renderProps": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Fetching",
  },
}
`;

exports[`data-loader multiple components load data once when props change 1`] = `
Object {
  "loadAllCompletedCalled": 2,
  "loadDataCount": 2,
  "renderCount": 4,
  "renderCount2": 3,
  "renderProps1": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Success2!",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
  "renderProps2": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Fetching",
  },
}
`;

exports[`data-loader resource can resolve synchronously 1`] = `"<div>Result!</div>"`;

exports[`data-loader resource can resolve synchronously when resource changes 1`] = `
Object {
  "events": Array [
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "59edd2a8": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Result!",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "fetch",
          },
          "status": "Idle",
        },
      },
      "type": "state-changed",
    },
    Object {
      "state": Object {},
      "type": "state-changed",
    },
    Object {
      "state": Object {
        "44f8a95a": Object {
          "data": Object {
            "dataFromServerSideRender": false,
            "hasData": true,
            "result": Object {
              "result": "Result2!",
            },
          },
          "lastAction": Object {
            "success": true,
            "type": "fetch",
          },
          "status": "Idle",
        },
      },
      "type": "state-changed",
    },
  ],
  "loadAllCompletedCalled": 0,
  "loadDataCount": 2,
  "renderActions": Object {
    "refresh": [Function],
  },
  "renderCount": 2,
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Object {
        "result": "Result2!",
      },
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`data-loader supports multiple loaders using the same key when data loading 1`] = `
Object {
  "loadAllCompletedCalled": 0,
  "loadDataCount": 1,
  "renderCount": 1,
  "renderCount2": 1,
  "renderProps1": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Fetching",
  },
  "renderProps2": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Fetching",
  },
}
`;
